### PR TITLE
Issue a warning if we detect a .js/.ts file collision

### PR DESCRIPTION
**Pull request by dfreeman**
This PR will cause a warning to be emitted if we detect the situation described in #780. Because of exactly the situation we're detecting, we can't tell _where_ the two conflicting files are coming from, but in most cases it will be apparent from context (e.g. the host app overriding a service from a particular addon).

Trying to describe the full situation and its possible ramifications would produce a giant wall of text in the terminal output, but the link in the warning points to #780, which has some good context and also links out to additional issues, like the upstream one in `broccoli-persistent-filter`.

image

### CODE REVIEW

**Comment by lifeart** [comment]
Issue happens if both files is `.ts` also

**Comment by dfreeman** [comment]
> also issue happens if both files is .ts

I don't think this is correct. Addons should never have `.ts` files in their `app/` trees, but if they do and the host also has a `.ts` file, it will override the addon's file.

**Comment by lifeart** [comment]
https://github.com/typed-ember/ember-cli-typescript/issues/724

**Comment by dfreeman** [comment]
Nothing in that thread indicates an issue when there were two conflicting `.ts` files when there wasn’t also a `.js` file of the same name floating around (e.g. because of differing addon versions that switched from including a `.ts` file to including a `.js` one)

If you can provide a reproduction where a nondeterministic build results from _only_ `.ts` files I’m happy to take a look, but if you consider the underlying problem here, that _can’t_ be caused by the same root issue as the one this PR is detecting, so it would need to be investigated and remedied separately.

File: ts/addon.ts [hunk 1]
```diff
@@ -97,6 +106,47 @@ export default addon({
...
+        if (path !== '/') return input;
+
+        let addon = this;
+        let checked = false;
+        let stew = require('broccoli-stew') as typeof import('broccoli-stew');
+
+        return stew.afterBuild(input, function() {
+          if (!checked) {
+            checked = true;
+            addon._checkForFileCollisions(this.inputPaths[0]);
```
Comment (chriskrycho) [inline ts/addon.ts]: Just checking: what invariant guarantees that `this.inputPaths` always has at least one element?

File: ts/tests/helpers/skeleton-app.ts [hunk 2]
```diff
@@ -44,7 +44,7 @@ export default class SkeletonApp {
     ));
   }
 
-  updatePackageJSON(callback: (arg: any) => string) {
+  updatePackageJSON(callback: (arg: any) => any) {
```
Comment (chriskrycho) [inline ts/tests/helpers/skeleton-app.ts]: 😭

Comment (dfreeman) [reply ts/tests/helpers/skeleton-app.ts]: Really this is `(arg: PackageJSON) => PackageJSON | void` — `string` was never right 🙃

File: ts/addon.ts (hunk 1 above)
Comment (dfreeman) [reply ts/addon.ts]: Broadly speaking, Broccoli plugins have one element in `inputPaths` per node they receive in their constructor.

In the case of `afterBuild` (and `Funnel` and a lot of other common plugins), it accepts a single node and then passes it in a singleton array when it calls `super` in its constructor, so `inputPaths[0]` in this case is the path on disk for the `input` tree we pass in a couple lines above.

Comment (chriskrycho) [reply ts/addon.ts]: Excellent; thank you! TIL!

**Comment by jamescdavis** [comment]
This has been released in [v3.1.3](https://github.com/typed-ember/ember-cli-typescript/releases/tag/v3.1.3).

**Comment by azhiv** [comment]
Can this warning be treated as an error during build? It would be nice to fail the build in case of a collision because otherwise proceeding with the collision (if missed) may result in significant harm.

**Pull request merged**